core/objects.go

```diff
@@ -252,6 +265,43 @@ type Challenge struct {
 
 	// Used by dns and dvsni challenges
 	Validation *jose.JsonWebSignature `json:"validation,omitempty"`
+
+	// Contains information about URLs used or redirected to and IPs resolved and
+	// used
+	ValidationRecord []ValidationRecord `json:"validationRecord,omitempty"`
+}
+
+// RecordsSane checks the sanity of a ValidationRecord object before sending it
+// back to the RA to be stored.
+func (ch Challenge) RecordsSane() bool {
+	if ch.ValidationRecord == nil || len(ch.ValidationRecord) == 0 {
+		return false
+	}
+
+	switch ch.Type {
+	case ChallengeTypeSimpleHTTP:
+		for _, rec := range ch.ValidationRecord {
+			if rec.URL == "" || rec.Hostname == "" || rec.Port == "" || rec.AddressUsed == nil ||
+				len(rec.AddressesResolved) == 0 {
+				return false
+			}
+		}
+	case ChallengeTypeDVSNI:
```

Comment:
Can we please stop doing these crazy cast statements with the json keys all in the same namespace? This is truly awful code to understand and is the kind of thing that will get long-term totally horked in our or other people's code. By using sub-object we get rid of the worrisome aspects of having the same names with differing expectations and prevent us from trying to "simplify" code by falling through in case statements. 

People will screw this up. We'll screw this up. It's super gross.